core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java

```diff
@@ -0,0 +1,262 @@
+                        + "\"type\":\"cluster_block_exception\","
+                        + "\"reason\":\"blocked by: [SERVICE_UNAVAILABLE/2/no master];\""
+                    + "}"
+                + "}"
+            + "},"
+            + "\"header\":{"
+                    + "\"foo_0\":\"0\","
+                    + "\"foo_1\":\"1\""
+                + "}"
+        + "}";
+
+        assertExceptionAsJson(e, false, equalTo(expectedJson));
+
+        ElasticsearchException parsed;
+        try (XContentParser parser = XContentType.JSON.xContent().createParser(expectedJson)) {
+            assertEquals(XContentParser.Token.START_OBJECT, parser.nextToken());
+            parsed = ElasticsearchException.fromXContent(parser);
+            assertEquals(XContentParser.Token.END_OBJECT, parser.currentToken());
+            assertNull(parser.nextToken());
+        }
+
+        assertNotNull(parsed);
+        assertThat(parsed.getMessage(), equalTo("Elasticsearch exception [type=exception, reason=foo]"));
+        assertThat(parsed.getHeaderKeys(), hasSize(4));
+        assertThat(parsed.getHeader("header_foo_0").get(0), equalTo("foo_0"));
+        assertThat(parsed.getHeader("header_foo_1").get(0), equalTo("foo_1"));
+        assertThat(parsed.getHeader("foo_0").get(0), equalTo("0"));
+        assertThat(parsed.getHeader("foo_1").get(0), equalTo("1"));
+
+        ElasticsearchException cause = (ElasticsearchException) parsed.getCause();
+        assertThat(cause.getMessage(), equalTo("Elasticsearch exception [type=exception, reason=bar]"));
+
+        cause = (ElasticsearchException) cause.getCause();
+        assertThat(cause.getMessage(), equalTo("Elasticsearch exception [type=exception, reason=baz]"));
+
+        cause = (ElasticsearchException) cause.getCause();
+        assertThat(cause.getMessage(),
+                equalTo("Elasticsearch exception [type=cluster_block_exception, reason=blocked by: [SERVICE_UNAVAILABLE/2/no master];]"));
+    }
+
+    public void testFromXContent() throws IOException {
+        final XContent xContent = randomFrom(XContentType.values()).xContent();
+        XContentBuilder builder = XContentBuilder.builder(xContent)
+                                                    .startObject()
+                                                        .field("type", "foo")
+                                                        .field("reason", "something went wrong")
+                                                        .field("stack_trace", "...")
+                                                    .endObject();
+
+        ElasticsearchException parsed;
+        try (XContentParser parser = xContent.createParser(builder.bytes())) {
+            assertEquals(XContentParser.Token.START_OBJECT, parser.nextToken());
+            parsed = ElasticsearchException.fromXContent(parser);
+            assertEquals(XContentParser.Token.END_OBJECT, parser.currentToken());
+            assertNull(parser.nextToken());
+        }
```

Comment:
That's a good suggestion but the content will be slightly different: when the content is parsed, it is encapsulated in a `ElasticsearchException`. 

It means that `{"type":"foo","reason":"something went wrong","stack_trace":"..."}` becomes a `ElasticsearchException` with message "Elasticsearch exception [type=foo, reason=something...]".

And such exception transformed again in XContent will be:
`{"type":"exception","reason":"Elasticsearch exception [type=foo, reason=something went wrong, stack_trace=...]"}`